plugins/ltac/profile_ltac.ml

```diff
@@ -388,8 +388,10 @@ let reset_profile () =
 (* ******************** *)
 
 let print_results_filter ~cutoff ~filter =
-  let valid id _ = Stm.state_of_id id <> `Expired in
-  data := SM.filter valid !data;
+  (* EJGA: This was quite strange, what is the layering here? How is
+     the tactic layer supposed to access the document one? *)
+  (* let valid id _ = Stm.state_of_id id <> `Expired in *)
+  (* data := SM.filter valid !data; *)
```

Comment:
IIRC the ltac profiler, in master, gathers all feedback messages containing profiling infos from the workers.
I think the variable is global, hence at the end of the day it may contain outdated data that is then filtered out.
now I don't recall exactly why it is done that way, but this code is only working fine in batch mode.

Reply:
Umm, I guess the right thing to do here is for `Profile_ltac` to install a `state_changed` listener and do its bookkeeping that way.

Reply:
The whole `Expired` stuff seems a bit strange, basically it should never be thrown in `master`, right? So I don't see what the check is needed, as long as you don't run that on the workers?

Reply:
IIRC, the check is needed to have backtracking work correctly.  In particular, you can't sync `data` with the document state, because that will erase profiling data at `Qed`.  So instead we need to do our own filtering.

Reply:
Umm, I don't see how `data` will be erased at Qed; it won't.

The problem seems more like there is not state synchronization between master and the workers, and lack of any sensible API forces to replicate STM functionality there.

But even so the current solution seems fragile, if the intended meaning of `print_results` is to print the data accumulated until that point of the document, the correct strategy wouldn't be to filter by expired but to filter for states less than the actual one.

Reply:
Currently, it won't drop data.  If we wanted to drop filtering and let the document sync mechanism keep track of it by making it a `Summary.ref`, it would, unless you've dropped the replay semantics for side effects at Qed.

> the correct strategy wouldn't be to filter by expired but to filter for spans that appear previously in the document.

Can you give an example of where these give different results?  Also how to filter for previous spans?  I can't actually speak much to the design decision of what to filter on.

Reply:
Yup, I see, you want profile data to "pierce" opacity. It seems we have no API for plugins to hook up above the "document" layer properly; using feedback kind of works [even if a bit hacky as it is really meant for UIs], but then, we need to extend the UI layer to reflect the new API, which we were planning to do later on, as it will be a protocol breaking change, and of all things, we need first to fix the `add` API so clients are not exposed to "ghost" spans!

I dunno how to proceed; maybe we could indeed extend the feedback internally in this PR, and keep `ide_slave` as is now. @gares 

> Can you give an example of where these give different results?

Umm, maybe not after all with the current implementation, but the current strategy depends a lot on the order things are being executed and in particular in how the cache is invalidated.

Reply:
One thing I'm taking from this thread is that these lines definitely deserve a better comment explaining what's going on.

Reply:
Ok, comments added, filtering restored by extending feedback as to include the document id.